### PR TITLE
Adding Promote/Demote users from/to guest endpoints

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Add the Promote/Demote API endpoint, App functions and store functions.

#### Ticket Link
[MM-15085](https://mattermost.atlassian.net/browse/MM-15085)

### CODE REVIEW

File: store/sqlstore/user_store.go [hunk 1]
```diff
@@ -1661,3 +1661,164 @@ func applyViewRestrictionsFilter(query sq.SelectBuilder, restrictions *model.Vie
...
+		}
+	}
+
+	query := sq.Update("Users").
+		Set("Roles", strings.Join(roles, " ")).
+		Where(sq.Eq{"Id": userId})
+
+	if us.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		query = query.PlaceholderFormat(sq.Dollar)
+	}
```
Comment (mkraft) [inline store/sqlstore/user_store.go]: Doesn't this get done automatically if you use `Where("Users.Id = ?", userId)`?

File: store/sqlstore/user_store.go [hunk 2]
```diff
@@ -1661,3 +1661,164 @@ func applyViewRestrictionsFilter(query sq.SelectBuilder, restrictions *model.Vie
...
+
+	query = sq.Update("TeamMembers").
+		Set("SchemeUser", true).
+		Set("SchemeGuest", false).
+		Where(sq.Eq{"UserId": userId})
+
+	if us.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		query = query.PlaceholderFormat(sq.Dollar)
+	}
```
Comment (mkraft) [inline store/sqlstore/user_store.go]: Same question here.

File: api4/user.go [hunk 3]
```diff
@@ -1829,3 +1831,73 @@ func getUserTermsOfService(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	user, err := c.App.GetUser(c.Params.UserId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	if !user.IsGuest() {
+		c.Err = model.NewAppError("Api4.promoteGuestToUser", "api.user.promote_guest_to_user.no_guest.app_error", nil, "", http.StatusNotImplemented)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (jespino) [reply api4/user.go]: It is what is expected here, If you are trying to promote a guest to a regular user, and the provided user, is not guest (is already a regular user) I expect that it fails. I think is more explicit than accept any user and expect that the conversion works.

File: api4/user.go [hunk 4]
```diff
@@ -1829,3 +1831,73 @@ func getUserTermsOfService(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	user, err := c.App.GetUser(c.Params.UserId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	if user.IsGuest() {
+		c.Err = model.NewAppError("Api4.demoteUserToGuest", "api.user.demote_user_to_guest.already_guest.app_error", nil, "", http.StatusNotImplemented)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (jespino) [reply api4/user.go]: Same reasoning here.

File: store/sqlstore/user_store.go [hunk 5]
```diff
@@ -1661,3 +1661,164 @@ func applyViewRestrictionsFilter(query sq.SelectBuilder, restrictions *model.Vie
 
 	return resultQuery
 }
+
+func (us SqlUserStore) PromoteGuestToUser(userId string) *model.AppError {
+	transaction, err := us.GetMaster().Begin()
+	if err != nil {
+		return model.NewAppError("SqlUserStore.PromoteGuestToUser", "store.sql_user.promote_guest.open_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
```
[the remark this answers is not in the document]
Comment (jespino) [reply store/sqlstore/user_store.go]: I agree with that, It is not necessary here, but is the pattern followed in the application I think. We can decide how we handle better the errors and that means the error translations too.

File: store/sqlstore/user_store.go (hunk 1 above)
Comment (jespino) [reply store/sqlstore/user_store.go]: nope, you have to explicitly setup this configuration to make it work with postgres.

File: store/sqlstore/user_store.go (hunk 2 above)
Comment (jespino) [reply store/sqlstore/user_store.go]: same answer :)

File: store/sqlstore/user_store.go (hunk 1 above)
Comment (mkraft) [reply store/sqlstore/user_store.go]: We're doing it like I suggested here, for example: https://github.com/mattermost/mattermost-server/pull/10701/files#diff-0b7dcefeb566578492d6be5ea6b2453cR898. Or maybe I'm misunderstanding?

Comment (jespino) [reply store/sqlstore/user_store.go]: ok, the difference is you are using `getQueryBuilder`at the beginning, and I'm using directly squirrel library, I think can change that, let me check.

Comment (jespino) [reply store/sqlstore/user_store.go]: Done! :)

File: api4/user.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkraft) [reply api4/user.go]: Depends, I suppose, on how you want your API to behave. *If it's significant to know* that the user was already a guest then this makes sense. Some examples where code doesn't throw an error when things are in the desired state:

* a SQL `UPDATE` query where the existing value is already same as the update value
* `chmod` where the existing file already has the update permissions

0/5 for me.

File: store/sqlstore/user_store.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkraft) [reply store/sqlstore/user_store.go]: A couple of lines down you're using a generic translation: https://github.com/mattermost/mattermost-server/pull/10791/files/08edce967c81a07c34a0dd161fc87f94849fee6f#diff-15eb08ffdad5b3c340cca5330dc9e5e4R1695.

File: api4/user.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jespino) [reply api4/user.go]: Yes, but in this case you are not establishing a value for the role of the user, you are "promoting a guest to user", which is a very clear an specific action executed on a kind of user. What I mean with this is, if I promote a guest to user and I pass an "admin" user, that should convert my admin user into a regular user?

File: store/sqlstore/user_store.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jespino) [reply store/sqlstore/user_store.go]: Yes, I get your point, but the problem is. Is this the place to solve the problem? or is it a project wide inconsistency that must be addressed in a more wide scope? I don't see the point of fixing it here, and anyway, I don't know which would be the proper fix here.

File: store/storetest/user_store.go [hunk 6]
```diff
@@ -3690,3 +3692,577 @@ func testUserStoreGetChannelGroupUsers(t *testing.T, ss store.Store) {
...
+		require.Nil(t, res.Err)
+		user := res.Data.(*model.User)
+
+		err := ss.User().PromoteGuestToUser(user.Id)
+		assert.Nil(t, err)
+		updatedUser, err := ss.User().Get(user.Id)
+		assert.Nil(t, err)
+		require.Equal(t, "system_user", updatedUser.Roles)
+	})
+	t.Run("Must work with guest user with teams but no channels", func(t *testing.T) {
```
Comment (mgdelacroix) [inline store/storetest/user_store.go]: nit due OCD: some tests have a white line between them and some others don't :stuck_out_tongue_closed_eyes:

Comment (jespino) [reply store/storetest/user_store.go]: Fixed, thanks! :)

**Pull request merged**